src/frequenz/sdk/actor/data_sourcing/microgrid_api_source.py

```diff
@@ -0,0 +1,330 @@
+    ComponentMetricId.VOLTAGE_PHASE_1: lambda msg: msg.voltage_per_phase[0],
+    ComponentMetricId.VOLTAGE_PHASE_2: lambda msg: msg.voltage_per_phase[1],
+    ComponentMetricId.VOLTAGE_PHASE_3: lambda msg: msg.voltage_per_phase[2],
+}
+
+_BatteryDataMethods: Dict[ComponentMetricId, Callable[[BatteryData], float]] = {
+    ComponentMetricId.SOC: lambda msg: msg.soc,
+    ComponentMetricId.SOC_LOWER_BOUND: lambda msg: msg.soc_lower_bound,
+    ComponentMetricId.SOC_UPPER_BOUND: lambda msg: msg.soc_upper_bound,
+    ComponentMetricId.CAPACITY: lambda msg: msg.capacity,
+    ComponentMetricId.POWER_LOWER_BOUND: lambda msg: msg.power_lower_bound,
+    ComponentMetricId.POWER_UPPER_BOUND: lambda msg: msg.power_upper_bound,
+}
+
+_InverterDataMethods: Dict[ComponentMetricId, Callable[[InverterData], float]] = {
+    ComponentMetricId.ACTIVE_POWER: lambda msg: msg.active_power,
+    ComponentMetricId.ACTIVE_POWER_LOWER_BOUND: lambda msg: msg.active_power_lower_bound,
+    ComponentMetricId.ACTIVE_POWER_UPPER_BOUND: lambda msg: msg.active_power_upper_bound,
+}
+
+_EVChargerDataMethods: Dict[ComponentMetricId, Callable[[EVChargerData], float]] = {
+    ComponentMetricId.ACTIVE_POWER: lambda msg: msg.active_power,
+    ComponentMetricId.CURRENT_PHASE_1: lambda msg: msg.current_per_phase[0],
+    ComponentMetricId.CURRENT_PHASE_2: lambda msg: msg.current_per_phase[1],
+    ComponentMetricId.CURRENT_PHASE_3: lambda msg: msg.current_per_phase[2],
+    ComponentMetricId.VOLTAGE_PHASE_1: lambda msg: msg.voltage_per_phase[0],
+    ComponentMetricId.VOLTAGE_PHASE_2: lambda msg: msg.voltage_per_phase[1],
+    ComponentMetricId.VOLTAGE_PHASE_3: lambda msg: msg.voltage_per_phase[2],
+}
+
+
+class MicrogridApiSource:
+    """Fetches requested metrics from the Microgrid API.
+
+    Used by the DataSourcingActor.
+    """
+
+    def __init__(
+        self,
+        registry: ChannelRegistry,
+    ) -> None:
+        """Create a `MicrogridApiSource` instance.
+
+        Args:
+            registry: A channel registry.  To be replaced by a singleton
+                instance.
+        """
+        self.comp_data_receivers: Dict[int, Receiver[Any]] = {}
+        self.comp_data_tasks: Dict[int, asyncio.Task[None]] = {}
+        self._registry = registry
+        self._req_streaming_metrics: Dict[
+            int, Dict[ComponentMetricId, List[ComponentMetricRequest]]
+        ] = {}
+
+    async def _check_battery_request(
+        self,
+        comp_id: int,
+        requests: Dict[ComponentMetricId, List[ComponentMetricRequest]],
+    ) -> None:
+        """Check if the requests are valid Battery metrics.
+
+        Raises:
+            ValueError: if the requested metric is not available for batteries.
+
+        Args:
+            comp_id: The id of the requested component.
+            requests: A list of metric requests received from external actors
+                for the given battery.
+        """
+        for metric in requests.keys():
```

Comment:
Nitpick: this is a bit redundant, `x in dict` is looking if `x` is in the keys of the dict already :)

```suggestion
        for metric in requests:
```

Reply:
This is one of the stupid things about python that lead to confusing bugs for me.  If the language always insists on `.keys()`,  that would have been amazing.  But I will remove it, to be consistent.

Reply:
done